### PR TITLE
v4.0.x: osc/ucx: Fix possible win creation/destruction race condition

### PR DESCRIPTION
**Pull request by janjust**
To avoid fully initializing the osc/ucx component for MPI application
that are not using One-Sided functionality, the initialization happens
at the first MPI window creation.

This commit ensures atomicity of global state modifications.

ported from: 6678ac0f557935b291ec2310216b7ea46e0c13b1
Signed-off-by: Artem Polyakov <artpol84@gmail.com>

### CODE REVIEW

File: ompi/mca/osc/ucx/osc_ucx_component.c [hunk 1]
```diff
@@ -21,6 +21,20 @@
     memcpy(((char*)(_dst)) + (_off), _src, _len); \
     (_off) += (_len);
 
+opal_mutex_t mca_osc_service_mutex = OPAL_MUTEX_STATIC_INIT;
+static void _osc_ucx_init_lock(void)
+{
+        if(mca_osc_ucx_component.enable_mpi_threads) {
+                    opal_mutex_lock(&mca_osc_service_mutex);
+                        }
```
Comment (hoopoepg) [inline ompi/mca/osc/ucx/osc_ucx_component.c]: alignment is really strange

File: ompi/mca/osc/ucx/osc_ucx_component.c [hunk 2]
```diff
@@ -349,7 +371,7 @@ static int component_select(struct ompi_win_t *win, void **base, size_t size, in
                                    &(mca_osc_ucx_component.ucp_worker));
         if (UCS_OK != status) {
             OSC_UCX_VERBOSE(1, "ucp_worker_create failed: %d", status);
-            ret = OMPI_ERROR;
+            ret = OMPI_ERR_TEMP_OUT_OF_RESOURCE;
```
Comment (brminich) [inline ompi/mca/osc/ucx/osc_ucx_component.c]: looks like `_osc_ucx_init_unlock()` will not be called in this case (and a few error cases below)

**Comment by artpol84** [comment]
@brminich, @hoopoepg good catch, thanks!

@janjust I've just revisited what we have in the master and it appears that there is one case that is not properly held there as well:

File: ompi/mca/osc/ucx/osc_ucx_component.c [hunk 3]
```diff
@@ -349,7 +371,7 @@ static int component_select(struct ompi_win_t *win, void **base, size_t size, in
...
             goto error_nomem;
```
Comment (artpol84) [inline ompi/mca/osc/ucx/osc_ucx_component.c]:
```suggestion
            goto select_unlock;
```

File: ompi/mca/osc/ucx/osc_ucx_component.c [hunk 4]
```diff
@@ -358,20 +380,44 @@ static int component_select(struct ompi_win_t *win, void **base, size_t size, in
...
         }
 
         if (mca_osc_ucx_component.enable_mpi_threads == true &&
             worker_attr.thread_mode != UCS_THREAD_MODE_MULTI) {
             OSC_UCX_VERBOSE(1, "ucx does not support multithreading");
-            ret = OMPI_ERROR;
+            ret = OMPI_ERR_TEMP_OUT_OF_RESOURCE;
             goto error_nomem;
```
Comment (artpol84) [inline ompi/mca/osc/ucx/osc_ucx_component.c]:
```suggestion
            goto select_unlock;
```

File: ompi/mca/osc/ucx/osc_ucx_component.c [hunk 5]
```diff
@@ -358,20 +380,44 @@ static int component_select(struct ompi_win_t *win, void **base, size_t size, in
         status = ucp_worker_query(mca_osc_ucx_component.ucp_worker, &worker_attr);
         if (UCS_OK != status) {
             OSC_UCX_VERBOSE(1, "ucp_worker_query failed: %d", status);
-            ret = OMPI_ERROR;
+            ret = OMPI_ERR_TEMP_OUT_OF_RESOURCE;
             goto error_nomem;
```
Comment (artpol84) [inline ompi/mca/osc/ucx/osc_ucx_component.c]:
```suggestion
            goto select_unlock;
```

File: ompi/mca/osc/ucx/osc_ucx_component.c (hunk 1 above)
Comment (janjust) [reply ompi/mca/osc/ucx/osc_ucx_component.c]: @hoopoepg thanks, and done, I forced pushed the changes; however. I'm not sure if that's allowed or if it required a new commit

File: ompi/mca/osc/ucx/osc_ucx_component.c (hunk 2 above)
Comment (janjust) [reply ompi/mca/osc/ucx/osc_ucx_component.c]: @brminich  thanks, and done, I forced pushed the changes; however. I'm not sure if that's allowed or if it required a new commit

File: ompi/mca/osc/ucx/osc_ucx_component.c [hunk 6]
```diff
@@ -349,29 +371,53 @@ static int component_select(struct ompi_win_t *win, void **base, size_t size, in
...
     }
+    
+    mca_osc_ucx_component.num_modules++;
+
+    OSC_UCX_ASSERT(mca_osc_ucx_component.num_modules > 0);
+    if (1 == mca_osc_ucx_component.num_modules) {
+        ret = opal_progress_register(progress_callback);
+        if (OMPI_SUCCESS != ret) {
+            OSC_UCX_VERBOSE(1, "opal_progress_register failed: %d", ret);
+            goto error;
```
Comment (brminich) [inline ompi/mca/osc/ucx/osc_ucx_component.c]: need to remove this, to unlock mutex

Comment (janjust) [reply ompi/mca/osc/ucx/osc_ucx_component.c]: @brminich thanks, sorry missed that one

Comment (brminich) [reply ompi/mca/osc/ucx/osc_ucx_component.c]: ok

**Comment by gpaulsen** [comment]
@brminich Please re-review after @janjust force-pushed.

File: ompi/mca/osc/ucx/osc_ucx_component.c (hunk 2 above)
Comment (artpol84) [reply ompi/mca/osc/ucx/osc_ucx_component.c]: @janjust that is the right way

**Pull request merged**